MmSupervisorPkg/Drivers/MmPeiLaunchers/MmDxeSupport.c

```diff
@@ -709,6 +712,276 @@ SmmIplSetVirtualAddressNotify (
   EfiConvertPointer (0x0, (VOID **)&mMmSupvCommunication.CommunicationRegion.VirtualStart);
 }
 
+// MM_SUPV: Update communicate buffer when entering DXE.
+
+/**
+  Communicate to MmSupervisor to update the buffer to runtime pages allocated in DXE.
+
+  @param[out] VersionInfo        Pointer to hold returned version information structure.
+
+  @retval EFI_SUCCESS            The version information was successfully queried.
+  @retval EFI_INVALID_PARAMETER  The VersionInfo was NULL.
+  @retval EFI_SECURITY_VIOLATION The Version returned by supervisor is invalid.
+  @retval Others                 Other error status returned during communication to supervisor.
+
+**/
+EFI_STATUS
+UpdateDxeCommunicateBuffer (
+  IN MM_SUPERVISOR_VERSION_INFO_BUFFER  *VersionInfo,
+  OUT MM_SUPERVISOR_COMM_UPDATE_BUFFER  *UpdatedCommBuffer
+  )
+{
+  UINTN                             Size;
+  EFI_STATUS                        Status;
+  MM_SUPERVISOR_REQUEST_HEADER      *RequestHeader;
+  MM_SUPERVISOR_COMM_UPDATE_BUFFER  *NewCommBuffer;
+  VOID                              *NewUserCommBuffer;
+  VOID                              *NewSupvCommBuffer;
+  VOID                              *NewMmCoreBuffer;
+
+  if ((VersionInfo == NULL) || (UpdatedCommBuffer == NULL)) {
+    ASSERT (FALSE);
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (VersionInfo->MaxSupervisorRequestLevel < MM_SUPERVISOR_REQUEST_COMM_UPDATE) {
+    // This means the supervisor is too old, cannot do such operation
+    return EFI_UNSUPPORTED;
+  }
+
+  // Now we are in the real deal, start with allocating new buffers
+  NewUserCommBuffer = AllocateAlignedRuntimePages (mMmUserCommonBufferPages, EFI_PAGE_SIZE);
+  NewSupvCommBuffer = AllocateAlignedRuntimePages (mMmSupvCommonBufferPages, EFI_PAGE_SIZE);
+  NewMmCoreBuffer   = AllocateAlignedRuntimePages (EFI_SIZE_TO_PAGES (sizeof (MM_CORE_PRIVATE_DATA)), EFI_PAGE_SIZE);
+
+  if ((NewUserCommBuffer == NULL) || (NewSupvCommBuffer == NULL)) {
+    return EFI_OUT_OF_RESOURCES;

```

Comment:
If the new comm buffer cannot be allocated and the PEI allocated comm buffer is boot services data, it will eventually cease to exist.

Right now the caller to this function prints an error debug message. Maybe at least an assert there would be useful to make sure attention is called to the failure in debug builds.

Reply:
Added a few asserts. Thanks